Design of issue tracking/reporting via GitHub (#1595)
dandi/dandi-archive: doc/design/dandiset-issues-tracking.md

- [ ] Setup https://github.com/dandisets [organization-wide](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file) [`.github/ISSUE_TEMPLATE`](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository#creating-issue-forms) with different templates/types of issues to report
- [ ] DLP: add a button "View/Report Issues" or "Issues" leading to issue Tracker (e.g. [Issues](https://github.com/dandisets/000108/issues) for 000108)
  - Complimentary/alternative: Could be "File an issue" button which would prefeed some issue body with `@dandiset-owner` handles etc.  But we might then need to introduce our own UI to choose among different types of issues
  - Optionally: dandi-archive could query/cache number of known open issues and include within that button, e.g. have it [(1) Issues](https://github.com/dandisets/000108/issues)
  - Complimentary:  asset-level issues filing in UI context menu per file, would pre-fill in the asset path

> - [ ] Optional: provide consistent collection of labels to outline typical problems which might occur

Author: FTR, `/users` endpoint might be the one to get the associated folks to alert, e.g. https://api.dandiarchive.org/api/dandisets/000026/users/

Author: There was a comment during odin to be able to request additional metadata to be provided... I think we might want not only a label but a specific issue template "Extra metadata request" or alike possibly with a structured record to fill out.

Another aspect/kind of request could be a query for specific processing or extension of dataset (eg for neuroscope project etc)

Author: as for labels etc -- added corresponding issues and umbrella issue about the bot: https://github.com/dandi/dandisets/issues/360 .  I will resolve this subthread to ease review.

  - e.g., `validation-error`, `unable-to-load`, `io-error`, `missing-file`, `permission`, etc.

> 

Reviewer: ```suggestion
### Changes on dandi-archive Web UI

Request moved from #1643 

- Add a section to the web UI for each DANDI set for buttons, such as "See issues", "File an issue", "Ask a question" which routes to the corresponding Issues page on the GitHub repo for that DANDI set
- Buttons or side component could also report the number of questions asked about that DANDI set
```

  - footer has overall link to report issues for the archive (and incorrect: [PR to fix](https://github.com/dandi/dandi-archive/pull/1594))
- We generally refer people to [helpdesk discussions](https://github.com/dandi/helpdesk/discussions) as common place to report issues
- While preparing DataLad versions of the Dandisets and posting them on GitHub we do not disable `Issues` feature:
  - for some Dandisets, e.g. https://github.com/dandisets/000108/issues we already used GitHub issues
  - we do not anyhow automatically add/invite original uploaders to those dandisets.
    - note: ATM we do not collect audit information on who specifically uploaded any asset to reliably deduce github users to assign.
- The issue discussing this: https://github.com/dandi/dandi-archive/issues/863

## Proposed Process

**Overall**: Reuse existing Dandisets on GitHub (https://github.com/dandisets), and available for them GitHub's issue tracker.

- [x] Enable Issue tracking per Dandiset.
  - In the process of creating DataLad Dandisets on GitHub we already enable `Issues` functionality for all dandisets.
- [ ] Initiate documentation in handbook on managing issues of the dandisets

> - [ ] Upon creation of a DataLad Dandiset on GitHub:

Reviewer 2: This is a major concern for me: currently, DataLad is *independently* creating repositories that happen to be useful for purposes of reporting issues with the Dandisets in the production instance. I would much prefer if each DANDI Archive deployment handled this (much as it handles DOI creation now).

Author: rright, let's adjust 

```suggestion
- [ ] Upon creation of a Dandiset on dandi-archive:
  - Create (via Github REST API) a git repository on GitHub in a configured (per instance) organization (referring to as `{organization}` below)
```

  - Invite all Dandiset owners in Triage role to the dandiset. In Triage scope they can close, label, assign issues. The other actions (e.g. user management) would need to be implemented using [bot].
  - Email them pointing to documentation (item above)

> - [ ] Setup https://github.com/dandisets [organization-wide](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file) [`.github/ISSUE_TEMPLATE`](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository#creating-issue-forms) with different templates/types of issues to report

Reviewer 2: Do you mean that the [dandisets org](https://github.com/dandisets) would be used across all instances to store all issues? I think it would be better to let each instance configure the org in which it will store its issue repos.

- [ ] Optional: provide consistent collection of labels to outline typical problems which might occur
  - e.g., `validation-error`, `unable-to-load`, `io-error`, `missing-file`, `permission`, etc.
  - ref: https://github.com/dandi/dandisets/issues/361

### Alternatives considered

#### GitHub, but not inviting users and operating using `@handles`

Although sounded like a good idea at first, cumbersome to implement with consistent and/or desired behavior:

- Dandiset owners would really have no power to do anything, like close the issue etc, unless we implement all such actions (assignment, labeling, etc) via a [bot].
- To ensure that clicking on "New issue" would add needed `@handles` we would need to either
  - add action to monitor new issues to enhance description with the `@handles`, or
  - generate/update issue template per each dandiset with its own list of owners.

### Changes on dandi-archive Backend

- Provision configuration settings for
  - GitHub organization where to create repository for a new dandiset
  - GitHub token with sufficient permissions to create repository in that organization

> - If there is provisioned GitHub organization: Job creating a new dandiset entry would trigger a job to create a new repository.

Reviewer 2: This section addresses the concerns I mentioned above, but there seems to be some contradiction between sections that we should resolve.

Reviewer 2: With https://github.com/dandi/dandi-archive/pull/1595/commits/d8adab53e4f9c7875eeb3c1a8225dcb3fcd4d2ef this is now resolved.

    - Buttons or side component could also report the number of questions asked about that DANDI set

### Benefits

- Users already must have a GitHub account to
  - register/login to dandiarchive: we will reuse that same mechanism
  - submit support requests in helpdesk
- GitHub provides a versatile and customizable issue tracking system many of the users might already be familiar with
- In Triage mode dandisets owners would have powers to assign/close issues
- We will make it optional, so if no GitHub organization provisioned, no backend/UI actions/elements.

### Disadvantages

- not "integrated" within dandiarchive.org as issues (meta)data would not be contained within DANDI.
  - A possible mitigation: I guess we could collect mirror issues/comments etc from GitHub internally in the archive. There are tools which could even be used to help. E.g. @yarikoptic has experience with using https://github.com/MichaelMure/git-bug to sync all issues from GitHub locally to collect all contributors to the project. E.g. [this script](https://github.com/nipy/heudiconv-joss-paper/blob/main/authors/tools/make-summaries#L92) processes a JSON dump of all issues from `git bug`  mirror.

> [bot]: https://github.com/dandi/dandisets/issues/360

Author: ```suggestion
[bot]: https://github.com/dandi/dandisets/issues/360

### Miscellaneous other possible future integrations with GitHub

- 2i2c uses github teams for resources management on the hub (ref: 1.1 within https://github.com/dandi/dandi-hub/pull/90/files#diff-82655098d9fb488babf6a5ce10d3d5f6a98d17b2f69de5ca28315e54e020bdf9R29)
```

- In Triage mode dandisets owners would have powers to assign/close issues
- We will make it optional, so if no GitHub organization provisioned, no backend/UI actions/elements.

### Disadvantages

- not "integrated" within dandiarchive.org as issues (meta)data would not be contained within DANDI.
  - A possible mitigation: I guess we could collect mirror issues/comments etc from GitHub internally in the archive. There are tools which could even be used to help. E.g. @yarikoptic has experience with using https://github.com/MichaelMure/git-bug to sync all issues from GitHub locally to collect all contributors to the project. E.g. [this script](https://github.com/nipy/heudiconv-joss-paper/blob/main/authors/tools/make-summaries#L92) processes a JSON dump of all issues from `git bug`  mirror.

[bot]: https://github.com/dandi/dandisets/issues/360

### Miscellaneous other possible future integrations with GitHub

- 2i2c uses github teams for resources management on the hub (ref: 1.1 within https://github.com/dandi/dandi-hub/pull/90/files#diff-82655098d9fb488babf6a5ce10d3d5f6a98d17b2f69de5ca28315e54e020bdf9R29)

> - Provide facility to propose changes to dandiset metadata:

Author: attn @bendichter on his request during DANDI/NWB meetup. Can resolve later
